File: python/backtracking/permutations.py
from typing import List

class Solution:
    def permutate(self, nums: List[int]) -> List[List[int]]:
        result = []

        def backtrack(path: List[int], used: List[bool]):
            if len(path) == len(nums):
                result.append(path[:])
                return

            for i in range(len(nums)):
                if used[i]:
                    continue


                used[i] = True

                path.append(nums[i])

                backtrack(path, used)

                path = path[:len(path)-1]

                used[i] = False


        backtrack([], [False] * len(nums))
        return result

File: golang/backtracking/permutations.go
package main

func permute(nums []int) [][]int {
	var result [][]int

	var backtrack func(path []int, used []bool)

	backtrack = func(path []int, used []bool) {
		if len(path) == len(nums) {
			permutation := make([]int, len(path))

			copy(permutation, path)

			result = append(result, permutation)

			return
		}

		for i := 0; i < len(nums); i++ {
			if used[i] {
				continue
			}

			used[i] = true

			path = append(path, nums[i])

			backtrack(path, used)

			path = path[:len(path)-1]

			used[i] = false
		}
	}

	backtrack([]int{}, make([]bool, len(nums)))
	return result
}
